docs/development/fleetlock/protocol.md

```diff
@@ -0,0 +1,95 @@
+---
+parent: Development
+nav_order: 3
+layout: default
+---
+
+# FleetLock protocol
+
+This document describes an HTTP-based protocol for orchestrating fleet-wide reboots, used by Zincati.
+It is modeled after a distributed counting semaphore with recursive locking and lock-ownership.
+
+## Overview
+
+This is a request-response protocol, where operations are always initiated by the client (i.e. Zincati).
+Each operation consists of a JSON payload sent as a POST request to the server.
+
+At an high level, the client can perform two operations:
+
+ * `RecursiveLock`: try to reserve (lock) a slot for rebooting
+ * `UnlockIfHeld`: try to release (unlock) a slot that it was previously holding
+
+Semaphore locks are owned, so that only the client that created a lock can release it. All operations are recursive, so that multiple unbalanced lock/unlock actions are allowed.
+
+## Client state-machine
+
+When a client starts it proceeds into either "initialization" or "finalization" state, based on system condition (i.e. whether a finalization is already in progress).
+In "**initialization**" state, the client tries to release any reboot slot it may have previously held.
+A successful unlock operation means that the client can proceed into its "**steady**" state and look for further updates.
+When an update is found and locally staged, the client proceed into its "**pre-reboot**" state and tries to lock a reboot slot.
+A successful lock operation means that the client can proceed into its "**finalization**" state and finalize a pending update, then reboot.
+
+## Requests
+
+### Endpoints
+
+All endpoints defined below are relative to a common deployment-specific base URL:
+
+ * `/v1/pre-reboot`: reserve/lock a reboot slot
+ * `/v1/steady-state`: release/unlock a reboot slot
+
+### Body
+
+All POST requests contain well-formed JSON body according to the following schema:
+
+ * `client_params` (object, mandatory)
+   * `id` (string, mandatory, non-empty): client identifier (e.g. node name or UUID)
+   * `group` (string, mandatory, non-empty): reboot-group of the client
+
+Client ID is a case-sensitive textual label that uniquely identifies a lock holder. It is generated and persisted by each client.
+Client group is a textual label conforming to the regexp `^[a-zA-Z0-9.-]+$`. The recommended default group name is "`default`".
```

Comment:
`group` also generated by the client right? Based on the Zincati config. Maybe this could be clearer. Will Zincati client ever send an empty group (requiring the server to do the defaulting?). Will it send a group that doesn't match the regex (i.e. is this info meant for the server implementer?).

Reply:
Yes, `group` is a client parameter in Zincati. It can be customized via configuration fragments. Zincati won't send a request with an empty or missing group value. Right now however it doesn't check validity with the regepx, that needs to be fixed, I've opened https://github.com/coreos/zincati/issues/351.

In general I would expect a server to only check the regexp and reject anything which does not conform. The "default" note is confusing here, so I'll drop that part.